### PR TITLE
Task/backend 47 setup socket server

### PR DESCRIPTION
**Pull request by WanderningMaster**

### CODE REVIEW

**Comment by WanderningMaster** [comment]
[Setup socket server](https://trello.com/c/0xvlMw1g/47-setup-socket-server)

File: .docker/backend.Dockerfile [hunk 1]
```diff
@@ -9,13 +9,14 @@ COPY ./package*.json ./
 COPY ./tsconfig.json ./
 COPY ./.eslintrc.yml ./
 COPY ./shared ./shared/
-COPY ./backend ./backend/
+COPY ./backend/core ./backend/core
 
-RUN npm ci -w shared -w backend
+RUN npm ci -w shared -w backend/core
+RUN npm run postinstall:backend
```
Comment (VitaliiBedletskyi) [inline .docker/backend.Dockerfile]: Why do you need manually run postinstall script?

File: .docker/docker-compose.local.yml [hunk 2]
```diff
@@ -4,12 +4,13 @@ services:
     image: rabbitmq:3-management-alpine
     hostname: rabbit
     networks:
+      - socket_network
```
Comment (VitaliiBedletskyi) [inline .docker/docker-compose.local.yml]: Why do you need a separate socket network?

File: backend/core/package.json [hunk 3]
```diff
@@ -16,11 +16,12 @@
     "postinstall": "prisma generate",
     "prestart": "npm run migrate:prod && npm run seed:soft",
     "start": "node ./build/server.js",
-    "build:ts": "tsc && tsc-alias -p tsconfig.json",
+    "build:ts": "tsc && tsc-alias -p ./tsconfig.json",
     "build": "npm run build:ts"
   },
   "devDependencies": {
     "@faker-js/faker": "7.3.0",
+    "@prisma/client": "^4.1.0",
```
Comment (VitaliiBedletskyi) [inline backend/core/package.json]: B19. The versions of dependencies used are fixed in package.json.

File: backend/core/package.json [hunk 4]
```diff
@@ -37,7 +38,6 @@
     "tsconfig-paths": "3.14.1"
   },
   "dependencies": {
-    "@prisma/client": "4.1.1",
```
Comment (VitaliiBedletskyi) [inline backend/core/package.json]: Why do you move @prisma/client to devDependencies?

File: backend/push/package.json [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  "dependencies": {
+    "amqplib": "0.10.2",
+    "body-parser": "1.20.0",
+    "dotenv": "16.0.1",
+    "express": "^4.18.1",
+    "glob": "8.0.3",
+    "helmet": "5.1.1",
+    "pino": "^8.4.0",
+    "pino-pretty": "^9.0.0",
+    "socket.io": "^4.5.1"
```
Comment (VitaliiBedletskyi) [inline backend/push/package.json]: B19. The versions of dependencies used are fixed in package.json.

File: backend/push/src/server.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import { app } from './app';
+import { CONFIG } from './config/config';
+import * as http from 'http';
+import { amqpService, socketService } from './services/services';
+import { logger } from './config/logger';
+
+const server: http.Server = http.createServer(app);
+socketService.subscribe(server);
+amqpService.connect();
```
Comment (VitaliiBedletskyi) [inline backend/push/src/server.ts]: Connect to the queue should be after starting the server.

File: package.json [hunk 7]
```diff
@@ -10,20 +10,24 @@
     "lint:css": "npm run lint:css --workspaces --if-present",
     "lint:js": "npm run lint:js --workspaces --if-present",
     "lint": "npm run lint:fs && npm run lint --workspaces --if-present",
-    "postinstall": "npm run build:shared && npm run postinstall -w backend",
+    "postinstall:backend": "npm run build:shared && npm run postinstall -w backend/core",
```
Comment (VitaliiBedletskyi) [inline package.json]: Why did you rename it? It's related not only to the backend app.

File: shared/src/common/enums/amqp/amqp-channel.enum.ts [hunk 8]
```diff
@@ -1,4 +1,6 @@
-export enum AmqpChannel {
+export enum AmqpQueue {
   STREAMLET = 'streamlet',
   STREAMLET_STUDIO = 'streamlet-studio',
+  SOCKET = 'socket',
+  SOCKET_ALL = 'socket-all',
```
Comment (VitaliiBedletskyi) [inline shared/src/common/enums/amqp/amqp-channel.enum.ts]: Use more specific names for queues

File: .docker/socket.Dockerfile [hunk 9]
```diff
@@ -0,0 +1,21 @@
+FROM node:16.16.0-slim as socket-build
```
Comment (VitaliiBedletskyi) [inline .docker/socket.Dockerfile]: Use more lightest node images like alpine

File: .docker/socket.Dockerfile [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+RUN apt-get update
+RUN apt-get install -y openssl
```
Comment (VitaliiBedletskyi) [inline .docker/socket.Dockerfile]: It's no needed for push server

**Review by VitaliiBedletskyi — CHANGES_REQUESTED** [review]
Does it seem it is an empty file? backend/push/src/messaging/ampq.ts

File: .docker/docker-compose.local.yml [hunk 11]
```diff
@@ -38,13 +39,35 @@ services:
...
       - backend_network
       - frontend_network
     logging:
       driver: "json-file"
       options:
         max-size: "10m"
         max-file: "5"
+  socket:
```
Comment (VitaliiBedletskyi) [inline .docker/docker-compose.local.yml]: Use the same name for your server, if you prefer a socket server name use it in all places...

File: .docker/backend.Dockerfile (hunk 1 above)
Comment (WanderningMaster) [reply .docker/backend.Dockerfile]: I`m not sure wht but when running postinstall in socket.Dockerfile it throws this error 
image

File: backend/core/package.json [hunk 12]
```diff
@@ -38,7 +38,7 @@
...
+    "@prisma/client": "^4.1.1",
```
Comment (VitaliiBedletskyi) [inline backend/core/package.json]: B19. The versions of dependencies used are fixed in package.json.

File: backend/push/src/server.ts [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+import { logger } from './config/logger';
+
+const server: http.Server = http.createServer(app);
+socketService.subscribe(server);
+
+server.listen(CONFIG.port, () => {
+  logger.info(`Server started on ${CONFIG.port} PORT`);
+});
+
+amqpService.connect();
```
Comment (VitaliiBedletskyi) [inline backend/push/src/server.ts]: The connection will be executed before start the express!!!

File: backend/push/src/config/amqp-connection.ts [hunk 14]
```diff
@@ -0,0 +1,15 @@
+import { CONFIG } from './config';
+import { connect, Channel } from 'amqplib';
+import { AmqpQueue } from '~/shared/enums/enums';
+
+const amqpConnect = async (): Promise<Channel> => {
+  const amqpServer = CONFIG.rabbitmqUrl;
+
+  const amqpChannel = await connect(amqpServer).then((connection) => connection.createChannel());
```
Comment (VitaliiBedletskyi) [inline backend/push/src/config/amqp-connection.ts]: I didn't see the logic to reconnect if the rabbitmq not started yet

**Review by VitaliiBedletskyi — CHANGES_REQUESTED** [review]

File: backend/push/src/config/amqp-connection.ts (hunk 14 above)
Comment (WanderningMaster) [reply backend/push/src/config/amqp-connection.ts]: I resolved it in docker-compose with 
restart: on-failure:5

Comment (VitaliiBedletskyi) [reply backend/push/src/config/amqp-connection.ts]: I don't recommend doing it this way.

File: backend/push/src/config/amqp-connection.ts [hunk 15]
```diff
@@ -0,0 +1,47 @@
+import { CONFIG } from './config';
+import { connect, Channel, Connection } from 'amqplib';
+import { AmqpQueue } from '~/shared/enums/enums';
+import { logger } from './logger';
+import { timeout } from '../helpers/timeout';
+
+const tryConnect = async (tries = 0): Promise<Connection> => {
+  const amqpServer = CONFIG.rabbitmqUrl;
+  if (tries >= 5) {
```
Comment (VitaliiBedletskyi) [inline backend/push/src/config/amqp-connection.ts]: Please not use magic numbers

File: backend/push/src/config/amqp-connection.ts [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+const amqpConnect = async (): Promise<Channel> => {
+  logger.info('Connecting to Rabbitmq...');
+  const connection: Connection = await tryConnect();
+
+  const amqpChannel = await connection.createChannel();
+  amqpChannel.assertQueue(AmqpQueue.NOTIFY_USER);
+  amqpChannel.assertQueue(AmqpQueue.NOTIFY_USER_BROADCAST);
+
+  return amqpChannel;
+  // return await connection.createChannel()
```
Comment (VitaliiBedletskyi) [inline backend/push/src/config/amqp-connection.ts]: Get rid of commented code

File: backend/push/src/config/amqp-connection.ts [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+  const amqpServer = CONFIG.rabbitmqUrl;
+  if (tries >= 5) {
+    logger.error('Connection failed');
+    process.exit(0);
+  }
+  try {
+    const connection = await connect(amqpServer);
+    connection.on('close', async () => {
+      logger.warn('Rabbitmq connection lost');
+      await timeout(3000);
```
Comment (VitaliiBedletskyi) [inline backend/push/src/config/amqp-connection.ts]: Better to usethe timeout with geometric progression step and start with 500ms

File: backend/push/src/config/amqp-connection.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+      return tryConnect(++tries);
+    });
+    logger.info('Connect to RabbitMQ success!');
+    return connection;
+  } catch {
+    logger.info('Trying to reconnect...');
+    await timeout(3000);
+    return tryConnect(++tries);
```
Comment (VitaliiBedletskyi) [inline backend/push/src/config/amqp-connection.ts]: Get rid of duplication

File: backend/push/src/config/amqp-connection.ts [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+    const connection = await connect(amqpServer);
+    connection.on('close', async () => {
+      logger.warn('Rabbitmq connection lost');
+      await timeout(3000);
+      return tryConnect(++tries);
+    });
+    logger.info('Connect to RabbitMQ success!');
+    return connection;
+  } catch {
+    logger.info('Trying to reconnect...');
```
Comment (VitaliiBedletskyi) [inline backend/push/src/config/amqp-connection.ts]: Please add try number in logs

File: backend/push/src/config/amqp-connection.ts [hunk 20]
```diff
@@ -0,0 +1,47 @@
+import { CONFIG } from './config';
+import { connect, Channel, Connection } from 'amqplib';
+import { AmqpQueue } from '~/shared/enums/enums';
+import { logger } from './logger';
+import { timeout } from '../helpers/timeout';
+
+const tryConnect = async (tries = 0): Promise<Connection> => {
+  const amqpServer = CONFIG.rabbitmqUrl;
+  if (tries >= 5) {
+    logger.error('Connection failed');
```
Comment (VitaliiBedletskyi) [inline backend/push/src/config/amqp-connection.ts]: Add more specific log error

File: backend/push/src/config/amqp-connection.ts [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+    process.exit(0);
```
Comment (VitaliiBedletskyi) [inline backend/push/src/config/amqp-connection.ts]: Please not use `process.exit(0)` in this place. You should throw an error and catch it in `backend/push/src/server.ts` file. See the example in `backend/core/server.ts` file

**Review by VitaliiBedletskyi — CHANGES_REQUESTED** [review]

File: backend/push/src/config/amqp-connection.ts (hunk 18 above)
Comment (WanderningMaster) [reply backend/push/src/config/amqp-connection.ts]: i dont get it how i can do that, move this code into function?

Comment (VitaliiBedletskyi) [reply backend/push/src/config/amqp-connection.ts]: Yes

**Review by VitaliiBedletskyi — APPROVED** [review]
